In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

df = pd.read_csv('/kaggle/input/house-prices-advanced-regression-techniques/train.csv')

# Предварительная обработка данных

1. Посмотреть на пропуски. 

In [33]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

In [34]:
df = pd.read_csv('/kaggle/input/house-prices-advanced-regression-techniques/train.csv')
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [35]:
missing_columns = df.columns[df.isnull().any()]
print(missing_columns)

Index(['LotFrontage', 'Alley', 'MasVnrType', 'MasVnrArea', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Electrical', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence',
       'MiscFeature'],
      dtype='object')


In [43]:
print(df.info())  # Сведения о типах данных и пропусках
print(df.describe())  # Статистическое описание числовых колонок

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1460 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          1460 non-null   object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

## Посмотрим на уникальные значения признаков с типом object
> *Let's look at the unique values of features with the object data type.*

In [37]:
object_columns = df.select_dtypes(include=['object'])

for column in object_columns.columns:
    unique_values = df[column].unique()
    print(f"Unique values in column '{column}':")
    print(unique_values)
    print('-' * 50)  # Separator for readability


Unique values in column 'MSZoning':
['RL' 'RM' 'C (all)' 'FV' 'RH']
--------------------------------------------------
Unique values in column 'Street':
['Pave' 'Grvl']
--------------------------------------------------
Unique values in column 'Alley':
[nan 'Grvl' 'Pave']
--------------------------------------------------
Unique values in column 'LotShape':
['Reg' 'IR1' 'IR2' 'IR3']
--------------------------------------------------
Unique values in column 'LandContour':
['Lvl' 'Bnk' 'Low' 'HLS']
--------------------------------------------------
Unique values in column 'Utilities':
['AllPub' 'NoSeWa']
--------------------------------------------------
Unique values in column 'LotConfig':
['Inside' 'FR2' 'Corner' 'CulDSac' 'FR3']
--------------------------------------------------
Unique values in column 'LandSlope':
['Gtl' 'Mod' 'Sev']
--------------------------------------------------
Unique values in column 'Neighborhood':
['CollgCr' 'Veenker' 'Crawfor' 'NoRidge' 'Mitchel' 'Somerst' 

## Обработаем числовые признаки

In [38]:
# Фасад участка: Линейный отрезок улицы, соединенный с объектом недвижимости
df['LotFrontage'] = df['LotFrontage'].fillna(df['LotFrontage'].median())

## Аккуратно обработаем объектные признаки

In [39]:
# Выход к аллее
df['Alley'] = df['Alley'].fillna('None')

**Обрати внимание**

25  MasVnrType     588 non-null    object 

26  MasVnrArea     1452 non-null   float64

**Для этих значение заполним модой**

In [44]:
# Вывод данных до замены
print('До замены', df.loc[df['MasVnrType'].isna() & (df['MasVnrArea'] > 0), ['MasVnrType', 'MasVnrArea']], sep='\n')

# Замена пропусков в MasVnrType на моду, если MasVnrArea > 0
df.loc[df['MasVnrType'].isna() & (df['MasVnrArea'] > 0), 'MasVnrType'] = df['MasVnrType'].mode()[0]

# Вывод данных после замены
print('После замены', df.loc[df['MasVnrType'].isna() & (df['MasVnrArea'] > 0), ['MasVnrType', 'MasVnrArea']], sep='\n')

До замены
     MasVnrType  MasVnrArea
624         NaN       288.0
773         NaN         1.0
1230        NaN         1.0
1300        NaN       344.0
1334        NaN       312.0
После замены
Empty DataFrame
Columns: [MasVnrType, MasVnrArea]
Index: []


**Для остальных None**

In [ ]:
# Заполнить пропуски специальной категорией 'None'
df['MasVnrType'].fillna('None', inplace=True)

**Обработаем камины**

56  Fireplaces     1460 non-null   int64  

57  FireplaceQu    770 non-null    object

Мы видим, что для 0 каминов FireplaceQu - Nan

In [42]:
# Вывод данных до замены
print('Before', df.loc[df['Fireplaces'] == 0, ['Fireplaces', 'FireplaceQu']].head(), sep='\n')

# Замена значений на 'None' там, где Fireplaces == 0
df.loc[df['Fireplaces'] == 0, 'FireplaceQu'] = df.loc[df['Fireplaces'] == 0, 'FireplaceQu'].fillna('None')

# Вывод данных после замены
print('After', df.loc[df['Fireplaces'] == 0, ['Fireplaces', 'FireplaceQu']].head(), sep='\n')


Before
    Fireplaces FireplaceQu
0            0         NaN
5            0         NaN
10           0         NaN
12           0         NaN
15           0         NaN
After
    Fireplaces FireplaceQu
0            0        None
5            0        None
10           0        None
12           0        None
15           0        None


In [32]:
df['FireplaceQu'].isnull().sum()

770

In [ ]:
df['MSZoning'].unique()

In [ ]:
df['PoolQC'].unique()
# titanic.Survived.unique()